yt_dlp/extractor/bilibili.py

```diff
@@ -590,14 +740,54 @@ def _real_extract(self, url):
                 duration=traverse_obj(initial_state, ('videoData', 'duration', {int_or_none})),
                 __post_extractor=self.extract_comments(aid))
         else:
-            return {
-                **metainfo,
-                'duration': float_or_none(play_info.get('timelength'), scale=1000),
-                'chapters': self._get_chapters(aid, cid),
-                'subtitles': self.extract_subtitles(video_id, cid),
-                'formats': self.extract_formats(play_info),
-                '__post_extractor': self.extract_comments(aid),
-            }
+            formats = self.extract_formats(play_info)
+
+            if not traverse_obj(play_info, ('dash')):  # for legacy-only formats
+                has_qn = lambda x: x in traverse_obj(formats, (..., 'quality'))
+                for qn in traverse_obj(play_info, ('accept_quality', lambda _, v: not has_qn(v), {int})):
+                    formats.extend(traverse_obj(
+                        self.extract_formats(self._download_playinfo(video_id, cid, headers=headers, qn=qn)),
+                        (lambda _, v: not has_qn(v.get('quality')))))
+                self.check_missing_formats(play_info, formats)
+                if traverse_obj(formats, lambda _, v: v['fragments']):
+                    if not self._configuration_arg('_prefer_multi_flv'):
+                        # `_prefer_multi_flv` is mainly for writing test case, user should hardly need this
+                        dropping = ', '.join(traverse_obj(formats, (
+                            lambda _, v: v['fragments'], {lambda x: f'{x["format"]} ({x["format_id"]})'})))
+                        formats = traverse_obj(formats, lambda _, v: not v.get('fragments'))
+                        if dropping:
+                            self.to_screen(f'Dropping incompatible flv format(s) {dropping} when mp4 exists')
+                    else:
+                        formats = traverse_obj(
+                            formats, lambda _, v: v['quality'] == int(self._configuration_arg('_prefer_multi_flv')[0]),
+                        ) or [max(traverse_obj(formats, lambda _, v: v['fragments']), key=lambda x: x['quality'])]
+
+            if formats[0].get('fragments'):  # transform multi_video format
```

Comment:
I'd add a comment saying that this was done since concatenation of flv files corrupts them.
I spent quite a while looking into the existing F4mFD code/flv parser, but couldn't get it to work. Since these are only legacy formats, I'd say the workaround of using multi video is fine.